examples/recording_service/pluggable_storage/cpp/FileStorageReader.hpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * (c) Copyright, Real-Time Innovations, 2018-.
+ * All rights reserved.
+ * No duplications, whole or partial, manual or electronic, may be made
+ * without express written permission.  Any such copies, or
+ * revisions thereof, must display this notice unaltered.
+ * This code contains trade secrets of Real-Time Innovations, Inc.
+ */
+
+#include "rti/recording/storage/StorageReader.hpp"
+
+#include <fstream>
+
+namespace cpp_example {
+
+/*
+ * Convenience macro to forward-declare the C-style function that will be
+ * called by RTI Recording Service to create your class.
+ */
+RTI_RECORDING_STORAGE_READER_CREATE_DECL(FileStorageReader);
+
+/**
+ * This class acts as a factory for objects of classes FileStorageStreamReader
+ * and FileStorageStreamInfoReader. These objects are used by Replay and/or
+ * Converter to retrieve data from the storage.
+ */
+class FileStorageReader :
+        public rti::recording::storage::StorageReader {
+public:
+    FileStorageReader(const rti::routing::PropertySet& properties);
+    virtual ~FileStorageReader();
+
+    virtual rti::recording::storage::StorageStreamInfoReader *
+            create_stream_info_reader(
+                    const rti::routing::PropertySet& properties);
+
+    virtual void delete_stream_info_reader(
+            rti::recording::storage::StorageStreamInfoReader *stream_info_reader);
+
+    virtual rti::recording::storage::StorageStreamReader * create_stream_reader(
+            const rti::routing::StreamInfo& stream_info,
+            const rti::routing::PropertySet& properties);
+
+    virtual void delete_stream_reader(
+            rti::recording::storage::StorageStreamReader *stream_reader);
+private:
+    std::ifstream info_file_;
+    std::ifstream data_file_;
+    std::string file_name_;
+};
+
+/*
+ * This class uses the provided DynamicData specialization of a
+ * StorageStreamReader to read data stored by the classes in file
+ * FileStorageWriter.hpp, for a given data stream (DDS topic).
+ * All data is stored in a text file. Data is serialized to a text format.
+ * Note: this example is only fit to work with the provided type definition
+ * (type HelloMsg - see HelloMsg.idl for more information).
+ */
+class FileStorageStreamReader :
+        public rti::recording::storage::DynamicDataStorageStreamReader {
+public:
+
+    FileStorageStreamReader(std::ifstream *data_file);
+
+    virtual ~FileStorageStreamReader();
```

Comment:
we don't need the virtual keyword in this class, since there is no further inheritance

Reply:
That's true. Will remove.